#
# Solaris specific patch to pass-through LANG and LC_ variables
# from the server environment to the child if they have not been
# accepted from the client.
#
--- old/session.c	2020-11-19 15:07:52.598515287 +0000
+++ new/session.c	2020-11-19 15:08:49.669098749 +0000
@@ -959,6 +959,39 @@ read_etc_default_login(char ***env, u_in
 }
 #endif /* HAVE_ETC_DEFAULT_LOGIN */
 
+
+/*
+ * Pass the given environment variable from the sshd environment through
+ * to the client only if it is not already present in env via SendEnv/AcceptEnv.
+ */
+static void
+pass_env(char **env, u_int *envsize, const char *name)
+{
+	char *server_value = getenv(name);
+
+	if (server_value != NULL && (options.num_accept_env == 0 ||
+	    child_get_env(env, name) == NULL)) {
+		child_set_env(&env, envsize, name, server_value);
+	}
+}
+/*
+ * Pass through the locale variables from the sshd environment
+ * if they have not already been accepted from the client via SendEnv/AcceptEnv.
+ */
+static void
+pass_locale_env(char **env, u_int *envsize)
+{
+	debug3("pass_locale_env");
+	pass_env(env, envsize, "LANG");
+	pass_env(env, envsize, "LC_ALL");
+	pass_env(env, envsize, "LC_CTYPE");
+	pass_env(env, envsize, "LC_COLLATE");
+	pass_env(env, envsize, "LC_TIME");
+	pass_env(env, envsize, "LC_NUMERIC");
+	pass_env(env, envsize, "LC_MONETARY");
+	pass_env(env, envsize, "LC_MESSAGES");
+}
+
 #if defined(USE_PAM) || defined(HAVE_CYGWIN)
 static void
 copy_environment_blacklist(char **source, char ***env, u_int *envsize,
@@ -1117,6 +1150,8 @@ do_setup_env(struct ssh *ssh, Session *s
 		    auth_sock_name);
 
 
+	pass_locale_env(env, &envsize);
+
 #ifdef PER_SESSION_XAUTHFILE
 	if (s->auth_file != NULL)
 		child_set_env(&env, &envsize, "XAUTHORITY", s->auth_file);
